common/types/test_util.go

```diff
@@ -0,0 +1,231 @@
+package types
+
+import (
+	"reflect"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/uber/cadence/common/testing/testdatagen"
+)
+
+type byteSizer interface{ ByteSize() uint64 }
+
+// AssertByteSizeMatchesReflect verifies v.(byteSizer).ByteSize()
+// matches an independent reflective deep-size computation.
+func AssertByteSizeMatchesReflect(t *testing.T, v any) {
+	t.Helper()
+
+	gen := testdatagen.New(t)
+
+	for i := 0; i < 100; i++ {
+
+		gen.Fuzz(v)
+
+		rv := reflect.ValueOf(v)
+		if !rv.IsValid() {
+			t.Fatalf("nil/invalid value")
+		}
+
+		// Must implement ByteSize (on value or pointer)
+		bs, ok := v.(byteSizer)
+		if !ok {
+			// if they passed a non-pointer, check its pointer receiver too
+			if rv.Kind() != reflect.Ptr {
+				if pv, ok2 := reflect.New(rv.Type()).Interface().(byteSizer); ok2 {
+					bs = pv
+					rv = rv.Addr()
+					ok = true
+				}
+			}
+		}
+		assert.True(t, ok, "%T does not implement ByteSize()", v)
+
+		// Compare impl vs reflect-based computation.
+		root := rv
+		if root.Kind() == reflect.Ptr {
+			root = root.Elem()
+		}
+		exp := uint64(root.Type().Size()) + structPayloadSize(root)
+		got := bs.ByteSize()
+
+		assert.Equal(t, exp, got, "ByteSize mismatch for %T: got=%d expected=%d (likely a new/changed field not reflected in ByteSize)", v, got, exp)
+	}
+}
+
+// AssertReachablesImplementByteSize ensures every reachable named struct
+// under root implements ByteSize (on value or pointer receiver).
+// You can pass a value (&T{}) or a type via (*T)(nil).
+func AssertReachablesImplementByteSize(t *testing.T, root any) {
+	t.Helper()
+
+	rt := reflect.TypeOf(root)
+
+	assert.NotNil(t, rt, "nil type")
+
+	if rt.Kind() == reflect.Ptr {
+		rt = rt.Elem()
+	}
+
+	seen := map[reflect.Type]bool{}
+	var missing []string
+	checkStructHasByteSize(rt, seen, &missing)
+
+	assert.Equal(t, 0, len(missing), "reachable struct types missing ByteSize(): %v", missing)
+}
+
+// ------------------- helpers -------------------
+
+func structPayloadSize(v reflect.Value) uint64 {
+	if v.Kind() == reflect.Ptr {
+		if v.IsNil() {
+			return 0
+		}
+		return typePayloadSize(v.Elem())
+	}
+	return typePayloadSize(v)
+}
+
+func typePayloadSize(v reflect.Value) uint64 {
+	if !v.IsValid() {
+		return 0
+	}
+
+	switch v.Kind() {
+	case reflect.Ptr:
+		if v.IsNil() {
+			return 0
+		}
+		// DO NOT call ByteSize() here — keep independence
```

Comment:
I would slightly expand on this comment so that it's clearer what you mean: you mean to ensure that the two means of calculating size are roughly equal and are not influencing one another?